### ISSUE TITLE
PageTitle doesn't display for SSR pages in mixed render mode app

### ISSUE THREAD

**Issue by marinasundstrom** [dotnet/aspnetcore#50268]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

**Info:** Investigation shows that this is related to enhanced navigation and DOM preservation. See comment.

I have a Blazor Server app with mixed render modes, SSR and WebAssembly (Auto). Some pages are server-rendered, and a couple of them have dynamic components (WebAssembly Auto) that set the title depending on content.

In ``App.razor``, I have added render mode to the ``HeadOutlet``:

```razor
<HeadOutlet @rendermode="@RenderMode.Auto" />
```

Upon navigating to to a page, ``PageTitle`` is then only updated/rendered for the pages with dynamic components setting them, but not the server-side rendered pages.

They instead default to their address:

<img width="160" alt="Screenshot 2023-08-22 at 20 26 52" src="https://github.com/dotnet/aspnetcore/assets/919580/433eea56-3af2-4f80-8317-d140f9070aa4">

The ```<title>``` element is not even rendered in ``<head></head>``.

### Expected Behavior

That PageTitle works in components of all modes a in mixed render mode app.

Even if ``HeadOutlet`` has the ``@rendermode="@RenderMode.Auto"`` the ``PageTitle``of SSR pages should be rendered.

### Steps To Reproduce

1. Create a Blazor Server app.
2. Update ``App.razor``:
```razor
<HeadOutlet @rendermode="@RenderMode.Auto" />
```

3. Create a page component which has a dynamic component that declares a ``PageTitle``.

### .NET Version

8.0.100-rc.2.23421.3

**Comment by marinasundstrom** [comment]
Upon investigating it further... This seems to be connected to the enhanced navigation and DOM preservation in the case of SSR. 

The ``title`` is in the request but isn't applied.

When you do a full navigation to the page the title changes.

**Comment by MackinnonBuck** [comment]
Yeah, this seems like another example of https://github.com/dotnet/aspnetcore/issues/49613 rearing its head again. I think we should aim for a fundamental fix for this that doesn't require any extra APIs or annotations.

**Comment by SteveSandersonMS** [comment]
@marinasundstrom @MackinnonBuck Are you sure this has something to do with DOM preservation or enhanced nav? I don't think it does, since the same issue will still occur even if you disable those via:

```js
Blazor.start({
    ssr: { disableDomPreservation: true }
});
```

Instead, I think what happens is:

 * `<HeadOutlet @rendermode=...>` prerenders, emitting the correct title into the response
 * Then the interactive runtime starts and re-runs `<HeadOutlet>`
 * But if the current page is *not* interactive, then there is no `<PageTitle>` anywhere in the circuit/runtime, and so `<HeadOutlet>` updates the page's title to nothing

So really I think this is an issue with interactive `HeadOutlet` believing that it should actively remove any existing title from the page if none was specified via `PageTitle`.

**Comment by marinasundstrom** [comment]
@SteveSandersonMS You are right. I was further looking into it an DOM preservation doesn't explain it. 

The problem is that the ```title`` gets removed when the render mode kicks in on a SSR page where ```PageTitle``` is not interactive. 

Doing a complete navigation hints about this.

**Comment by MackinnonBuck** [comment]
I've looked into this a bit more. `PageTitle` is actually working as designed. If you add a separate `<title>` element directly to the `<head>`, then that acts as the default title instead of "localhost...".

However, note that all interactive components immediately replace their prerendered content. If the `<HeadOutlet>` is interactive, its content will get replaced with that of any interactive `<HeadContent>` or `<PageTitle>`. Since there are no interactive `<HeadContent>` or `<PageTitle>` components on the page, the `<HeadOutlet>` renders the default content (which includes whatever the default title is).

For now, we would recommend that `<HeadContent>` be used without an interactive render mode to avoid this problem.

**Comment by marinasundstrom** [comment]
@MackinnonBuck Ok. I’ll have to find a workaround then.

My case was a product listing page. I wanted to title to reflect the state of the interactive component.

I can probably set the title from JavaScript.

**Comment by MackinnonBuck** [comment]
When addressing this issue, we should also consider the case where the `<PageTitle>` gets set via streaming update as reported in https://github.com/dotnet/aspnetcore/issues/52689.

**Comment by mkArtakMSFT** [comment]
Closing as a dupe of https://github.com/dotnet/aspnetcore/issues/51132

**Comment by SteveSandersonMS** [comment]
I'm reopening this issue because I don't think it's best understood as a duplicate of https://github.com/dotnet/aspnetcore/issues/51132

My reasoning is that there might not be a completely general solution for #51132 since not section usage will even make sense in a mixed rendermode scenario. And it's potentially complex enough that it would be expensive to address to any extent.

**But** just finding a standalone solution for PageTitle may be much more practical. If this is the main thing that's causing trouble for people, there's a good argument for addressing this limitation on its own.

**Comment by SteveSandersonMS** [comment]
@dotnet/aspnet-blazor-eng I've removed loads of labels, milestone, etc. so this goes back into triage.

**Comment by mkArtakMSFT** [comment]
Parking this in the same milestone as https://github.com/dotnet/aspnetcore/issues/51132. Let's see if we can actually solve the more general problem tracked by that issue, and if not, then try to solve the PageTitle case as a special one.

**Comment by marqdouj** [comment]
fwiw, I just started working LocalProtectedStorage so I was using this render mode and reading the store in OnIntializedAsync.

@rendermode @(new InteractiveServerRenderMode(prerender: false))

So I changed the rendermode to InteractiveServer and used OnAfterRenderAsync/FirstRender to read the store; PageTitle now works.

**Comment by kallebysantos** [comment]
Hi guys, I would like to share my workaround `<PageTitleAsync />` component:

<details>
<summary>code</summary>

*Page.razor*
```razor
@attribute [StreamRendering]

@if(IsLoading)
{
    @* you can also use the <PageTitle> at first time to get SEO <head> but it causes a small delay between title switching*@
    <PageTitleAsync>Loading.. | Profile</PageTitleAsync>
}
else if (User is not null)
{
     @*To explicit use `rendermode` you may need to pass content as Parameter*@
    <PageTitleAsync>@(User.Name) | Profile</PageTitleAsync>

   <h1>Hello @(User.Name)</h1>
 }
``` 

*PageTitleAsync.razor*
```razor
@inject IJSRuntime JS

<title @ref=@(_titleRef) class="sr-only hidden">
    @ChildContent
</title>

@code {
    [Parameter, EditorRequired]
    public required RenderFragment ChildContent { get; set; }

    private IJSObjectReference? _jsModule;

    private ElementReference _titleRef;

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            _jsModule = await JS.InvokeAsync<IJSObjectReference>(
                identifier: "import",
                args: "./Components/UI/PageTitle/SSRPageTitle.razor.js"
            );

            await _jsModule.InvokeVoidAsync("SetPageTitle", _titleRef);
        }
    }
}
``` 

*PageTitleAsync.razor.js*
```js
export function SetPageTitle(titleElement) {
  document.title = titleElement.innerText;
}
```
